### PR TITLE
Add positive testcases for config_id verification on icelake VMs

### PR DESCRIPTION
**Pull request by manojrupireddy**
Additional check is added to skip tests on coffee lake VMs.
Fix https://github.com/openenclave/openenclave/issues/3903

Made changes to oesign tool. Users need to specify EnableKSS field in conf file to leverage SGX KSS features.
Signed-off-by: manoj rupireddy <marupire@microsoft.com>

### CODE REVIEW

File: tests/config_id/enc/enc_kss.c [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+oe_result_t enclave_test_config_id()
+{
+    OE_TRACE_INFO("enclave_config_id_test_kss_properties invoked\n");
+    oe_result_t result = OE_UNEXPECTED;
+    size_t report_size = OE_MAX_REPORT_SIZE;
+    uint8_t* remote_report = NULL;
+    oe_report_header_t* header = NULL;
+    sgx_quote_t* quote = NULL;
+
+    result = oe_get_report(
```
Comment (yentsanglee) [inline tests/config_id/enc/enc_kss.c:19]: Please call the 'evidence' set of the API. I.e.

1. oe_get_evidence()
2. oe_verify_evidence()
3. _find_claim(OE_CLAIM_SGX_CONFIG_ID)
4. _find_claim(OE_CLAIM_SGX_CONFIG_SVN)
5. memcmp(configid)
6. memcmp(configsvn)

Similar check in openenclave\tests\tools\oesign\test-enclave\enclave\enc.c between line 180 and 240.

File: tests/config_id/enc/enc_kss.c [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+                sizeof(original_config_id)))
+        {
+            OE_TRACE_WARNING("========== Read wrong config id from the report");
+            result = OE_REPORT_PARSE_ERROR;
+        }
+        if (report_body->configsvn != original_config_svn)
+        {
+            OE_TRACE_WARNING(
+                "========== Read wrong config svn from the report");
+            result = OE_REPORT_PARSE_ERROR;
```
Comment (yentsanglee) [inline tests/config_id/enc/enc_kss.c:50]: Add OE_TRACE_INFO() to print out the retrieved config_id and configsvn.

**Review by mingweishih — COMMENTED** [review]
A high-level question: can we consolidate this test and the previous config_id test?

**Comment by manojrupireddy** [comment]
> A high-level question: can we consolidate this test and the previous config_id test?

I think no. This enclave is signed with KSS flags. The negative tests one is not signed with KSS flags. Also adding @yentsanglee here.

**Comment by mingweishih** [comment]
> > A high-level question: can we consolidate this test and the previous config_id test?
> 
> I think no. This enclave is signed with KSS flags. The negative tests one is not signed with KSS flags. Also adding @yentsanglee here.

At minimal, you can have use the host and add arguments to differentiate different enclaves (see https://github.com/openenclave/openenclave/tree/master/tests/debug-mode/host for an example).
By the way, what happened if we run a KSS-signed enclave on the SGX1 machine? What's the difference between this case and the existing config_id test?

**Comment by manojrupireddy** [comment]
> > > A high-level question: can we consolidate this test and the previous config_id test?
> > 
> > 
> > I think no. This enclave is signed with KSS flags. The negative tests one is not signed with KSS flags. Also adding @yentsanglee here.
> 
> At minimal, you can have use the host and add arguments to differentiate enclaves (see https://github.com/openenclave/openenclave/tree/master/tests/debug-mode/host for an example).
> By the way, what happened if we run a KSS-signed enclave on the SGX1 machine? What's the difference between this case and the existing config_id test?
> 
> Also, we're missing the case that the enclave is actually signed (with a config file).
The enclave creation fails if we want to sign image with KSS flags on SGX1 machine. [here](https://github.com/openenclave/openenclave/blob/master/host/sgx/create.c#L888)
The previous tests confirms that enclave creation fails if config_id & config_svn settings are passed as a mandatory ones on SGX1 machine . If config_id is set as optional one, enclave will be created successfully on SGX1 machine. These 2 options are tested as part of previous test cases. The new test case tests the path where config_id settings are made mandatory and  it is retrieved successfully once the enclave is created succesfully on icelake machine.

**Comment by mingweishih** [comment]
> > > > A high-level question: can we consolidate this test and the previous config_id test?
> > > 
> > > 
> > > I think no. This enclave is signed with KSS flags. The negative tests one is not signed with KSS flags. Also adding @yentsanglee here.
> > 
> > 
> > At minimal, you can have use the host and add arguments to differentiate enclaves (see https://github.com/openenclave/openenclave/tree/master/tests/debug-mode/host for an example).
> > By the way, what happened if we run a KSS-signed enclave on the SGX1 machine? What's the difference between this case and the existing config_id test?
> 
> Also, we're missing the case that the enclave is actually signed (with a config file).
> 
> The enclave creation fails if we want to sign image with KSS flags on SGX1 machine. [here](https://github.com/openenclave/openenclave/blob/master/host/sgx/create.c#L888)

So do we miss this negative test?

> The previous tests confirms that enclave creation fails if config_id & config_svn settings are passed as a mandatory ones on SGX1 machine . If config_id is set as optional one, enclave will be created successfully on SGX1 machine. These 2 options are tested as part of previous test cases. The new test case tests the path where config_id settings are made mandatory and it is retrieved successfully once the enclave is created succesfully on icelake machine.
> [kss flags signing is done here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/enc/enc_kss.c#L56)

`oesign` does not involve if we use the macro, which is for the debug purpose. You may also need to have a case that signs the enclave using `oesign`. To do so, add a conf file and specify the file with the `CONFIG` argument in the `add_enclave`. See https://github.com/openenclave/openenclave/blob/master/tests/debug-mode/enc/CMakeLists.txt#L28.

**Comment by manojrupireddy** [comment]
> > > > > A high-level question: can we consolidate this test and the previous config_id test?
> > > > 
> > > > 
> > > > I think no. This enclave is signed with KSS flags. The negative tests one is not signed with KSS flags. Also adding @yentsanglee here.
> > > 
> > > 
> > > At minimal, you can have use the host and add arguments to differentiate enclaves (see https://github.com/openenclave/openenclave/tree/master/tests/debug-mode/host for an example).
> > > By the way, what happened if we run a KSS-signed enclave on the SGX1 machine? What's the difference between this case and the existing config_id test?
> > 
> > 
> > Also, we're missing the case that the enclave is actually signed (with a config file).
> > The enclave creation fails if we want to sign image with KSS flags on SGX1 machine. [here](https://github.com/openenclave/openenclave/blob/master/host/sgx/create.c#L888)
> 
> So do we miss this negative test?
> 
> > The previous tests confirms that enclave creation fails if config_id & config_svn settings are passed as a mandatory ones on SGX1 machine . If config_id is set as optional one, enclave will be created successfully on SGX1 machine. These 2 options are tested as part of previous test cases. The new test case tests the path where config_id settings are made mandatory and it is retrieved successfully once the enclave is created succesfully on icelake machine.
> > [kss flags signing is done here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/enc/enc_kss.c#L56)
> 
> `oesign` does not involve if we use the macro, which is for the debug purpose. You may also need to have a case that signs the enclave using `oesign`. To do so, add a conf file and specify the file with the `CONFIG` argument in the `add_enclave`. See https://github.com/openenclave/openenclave/blob/master/tests/debug-mode/enc/CMakeLists.txt#L28.

Got it. I will check and add oesign case.  
To summarize existing test cases,
case 1: on SGX1 machine, if config_id is mandatory enclave creation should fail [here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/host/host.c#L67)
case 2: on SGX1 machine, if config_id is optional enclave creation should be successful [here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/host/host.c#L54)
case 3: on Icelake VM, if config_id is mandatory enclave creation should be successful and developer should be able to read back config_id on enclave side.  [here](if config_id is mandatory enclave creation should fail)

**Comment by mingweishih** [comment]
> Got it. I will check and add oesign case.
> To summarize existing test cases,
> case 1: on SGX1 machine, if config_id is mandatory, enclave creation should fail [here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/host/host.c#L67)
> case 2: on SGX1 machine, if config_id is optional, enclave creation should be successful [here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/host/host.c#L54)
> case 3: on Icelake VM, if config_id is mandatory, enclave creation should be successful and developer should be able to read back config_id on enclave side. [here](https://github.com/manojrupireddy/openenclave/blob/config_id_positive_tests/tests/config_id/host/host_kss.c#L63)

Can we have two more cases: set the as mandatory and optional and the enclave is signed with KSS? The mandatory case can be added to a `else` of the `if (_is_kss_supported())`.  The optional case should always succeed regardless the return of _is_kss_supported()

File: tests/config_id/enc/CMakeLists.txt [hunk 3]
```diff
@@ -10,14 +10,61 @@ add_custom_command(
     edger8r --trusted ${EDL_FILE} --search-path ${PROJECT_SOURCE_DIR}/include
     ${DEFINE_OE_SGX} --search-path ${CMAKE_CURRENT_SOURCE_DIR})
 
+add_enclave(
+  TARGET
+  config_id_kss_enc
+  UUID
```
Comment (mingweishih) [inline tests/config_id/enc/CMakeLists.txt:16]: You can remove the UUID as it's optee-specific.

File: tests/config_id/enc/sign.conf [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) Open Enclave SDK contributors.
+# Licensed under the MIT License.
+
+# Enclave settings:
+Debug=0
+NumHeapPages=1024
+NumStackPages=1024
+NumTCS=2
+ProductID=1
+SecurityVersion=1
```
Comment (mingweishih) [inline tests/config_id/enc/sign.conf:10]: extra line

File: tests/config_id/host/host.c [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+static bool _is_kss_supported()
+{
+    uint32_t eax, ebx, ecx, edx;
+    eax = ebx = ecx = edx = 0;
+
+    // Obtain feature information using CPUID
+    oe_get_cpuid(0x12, 0x1, &eax, &ebx, &ecx, &edx);
+
+    // Check if KSS (bit 7) is supported by the processor
+    if (!(eax & (1 << 7)))
```
Comment (mingweishih) [inline tests/config_id/host/host.c:19]: return (eax & (1 << 7));

File: tests/config_id/host/host.c [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+{
+    oe_result_t result;
+    oe_enclave_t* enclave = NULL;
+
+    if (argc < 3)
+    {
+        fprintf(
+            stderr,
+            "Usage: %s ENCLAVE_PATH [no_kss_flags|with_kss_flags] "
+            "[--host-threads n] [--enclave-threads n] "
```
Comment (mingweishih) [inline tests/config_id/host/host.c:35]: are these options relevant?

**Review by mingweishih — COMMENTED** [review]
Can you also test:
enclave with prop_kss.c + sign.conf?

File: tests/config_id/host/host.c (hunk 6 above)
Comment (manojrupireddy) [reply tests/config_id/host/host.c:35]: not sure about [--host-threads n] [--enclave-threads n]  ..I continued it from previous code

File: tests/config_id/enc/enc.c [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+        evidence,
+        evidence_size,
+        NULL,
+        0,
+        NULL,
+        0,
+        &claims,
+        &claims_length));
+
+    // OE_TRACE_ERROR("========== null here");
```
Comment (mingweishih) [inline tests/config_id/enc/enc.c:75]: Remove the commented line

File: tests/config_id/enc/enc.c [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    if (memcmp(config_id, original_config_id, sizeof(original_config_id)) ||
+        memcmp(config_svn, &original_config_svn, sizeof(original_config_svn)))
+    {
+        OE_TRACE_WARNING(
+            "========== Read wrong config id / config svn from the report");
+        result = OE_REPORT_PARSE_ERROR;
+    }
+
+done:
+    OE_CHECK(oe_free_evidence(evidence));
```
Comment (mingweishih) [inline tests/config_id/enc/enc.c:94]: `OE_CHECK` should be used prior to `done:`, otherwise it could lead to infinite loop.

File: tests/config_id/enc/sign_kss.conf [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+# Licensed under the MIT License.
+
+# Enclave settings:
+Debug=1
+NumHeapPages=1024
+NumStackPages=1024
+NumTCS=2
+ProductID=1
+SecurityVersion=1
+ExtendedProductID=25419627-14f6-4625-9329-cf5f10a57feb
```
Comment (mingweishih) [inline tests/config_id/enc/sign_kss.conf:11]: Are this the following configurations used anywhere?

File: tests/config_id/enc/props_kss.c [hunk 10]
```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <openenclave/enclave.h>
+
+OE_SET_ENCLAVE_SGX_KSS(
+    1, /* ProductID */
+    1, /* SecurityVersion */
+    {0},
```
Comment (mingweishih) [inline tests/config_id/enc/props_kss.c:9]: Can we set ExtendedProductID / ExtendedProductID here?

File: tests/config_id/enc/enc.c [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    uint8_t* config_id;
+    uint16_t* config_svn;
+
+    config_id =
+        (uint8_t*)_find_claim(claims, claims_length, OE_CLAIM_SGX_CONFIG_ID);
+
+    config_svn =
+        (uint16_t*)_find_claim(claims, claims_length, OE_CLAIM_SGX_CONFIG_SVN);
+
+    if (memcmp(config_id, original_config_id, sizeof(original_config_id)) ||
```
Comment (qiucwang) [inline tests/config_id/enc/enc.c:85]: Check the `config_id` and `config_svn` separately so we can know which value is not correct.

File: tests/config_id/enc/enc.c [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        memcmp(config_svn, &original_config_svn, sizeof(original_config_svn)))
+    {
+        OE_TRACE_WARNING(
+            "========== Read wrong config id / config svn from the report");
+        result = OE_REPORT_PARSE_ERROR;
```
Comment (qiucwang) [inline tests/config_id/enc/enc.c:90]:
```suggestion
        OE_RAISE_MSG(
            OE_REPORT_PARSE_ERROR,
            "========== Read wrong config svn(0x%x) from the report, expected(0x%x)", 
            *config_svn, 
            original_config_svn);
```

File: tools/oesign/oesign.c [hunk 13]
```diff
@@ -490,6 +490,9 @@ void _merge_config_file_options(
 
     if (options->family_id.has_value || options->extended_product_id.has_value)
         properties->config.attributes |= SGX_FLAGS_KSS;
+    else /* Disable SGX KSS Flags if family id and extended product id absent in
+            conf file */
```
Comment (mingweishih) [inline tools/oesign/oesign.c:494]:
```suggestion
            the configuration file */
```

File: tests/config_id/enc/enc.c (hunk 8 above)
Comment (manojrupireddy) [reply tests/config_id/enc/enc.c:94]: can I remove OE_CHECK s from here...should not the underlying function handle the exception scenarios.

File: tests/config_id/enc/sign_kss.conf (hunk 9 above)
Comment (manojrupireddy) [reply tests/config_id/enc/sign_kss.conf:11]: they are not used anywhere for config_id/config svn test purpose

File: tests/config_id/enc/props_kss.c (hunk 10 above)
Comment (manojrupireddy) [reply tests/config_id/enc/props_kss.c:9]: raised issue https://github.com/openenclave/openenclave/issues/4112

**Comment by manojrupireddy** [comment]
bors try

File: tests/config_id/enc/enc.c [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        (uint16_t*)_find_claim(claims, claims_length, OE_CLAIM_SGX_CONFIG_SVN);
+
+    if (memcmp(config_id, original_config_id, sizeof(original_config_id)))
+    {
+        OE_TRACE_INFO("\noriginal_config_id :\n");
+        for (size_t i = 0; i < sizeof(original_config_id); i++)
+            OE_TRACE_INFO("0x%0x ", (uint8_t)original_config_id[i]);
+        OE_TRACE_INFO("\nparsed config_id :\n");
+        for (size_t i = 0; i < sizeof(original_config_id); i++)
+            OE_TRACE_INFO("0x%0x ", (uint8_t)config_id[i]);
```
Comment (qiucwang) [inline tests/config_id/enc/enc.c:90]:
```suggestion
        OE_TRACE_INFO("\noriginal_config_id :\n0x");
        for (size_t i = 0; i < sizeof(original_config_id); i++)
            OE_TRACE_INFO("%0x", (uint8_t)original_config_id[i]);
        OE_TRACE_INFO("\nparsed config_id :\n0x");
        for (size_t i = 0; i < sizeof(original_config_id); i++)
            OE_TRACE_INFO("%0x", (uint8_t)config_id[i]);
```
So that byte outputs are not separated by "[TRACE INFO HEADER] 0x"

Comment (qiucwang) [inline tests/config_id/enc/enc.c:90]:
```suggestion
        OE_TRACE_INFO("\noriginal_config_id :\n0x");
        oe_hex_dump(original_config_id, OE_COUNT_OF(original_config_id));
        OE_TRACE_INFO("\nparsed config_id :\n0x");
        oe_hex_dump(config_id, OE_COUNT_OF(original_config_id));
```
So that each byte is not separate by OE_TRACE_INFO header.

**Review by qiucwang — APPROVED** [review]

**Comment by manojrupireddy** [comment]
With the latest revisions, I have introduced a new field called 'EnableKSS' in conf file. User need to set this flag to 1 to leverage SGX KSS features. This will have indirect impact on developers planning to leverage FamilyID and ExtendedProductID fields of SGX2 machine. In the present code version of oesign.c in the master branch, kss flag is enabled if familyID and ExtendedProductID are present in conf file. With my current code, Instead of oesign.c deducting the logic to enable kss flag or not based on presence/absence of familyID and ExtendedProductId, developers need to explicitly mention if they want to enable KSS or not. This additional field will be helpful in the scenario where developers want to use configid/configsvn fields but not familyId and ExtendedProductiD fields. @bodzhang @yentsanglee @radhikaj @mingweishih @anakrish  can you see if this new field will have any unforeseen implications.

File: tools/oesign/oesign.c [hunk 15]
```diff
@@ -488,7 +512,7 @@ void _merge_config_file_options(
             &options->extended_product_id.value,
             sizeof(options->extended_product_id.value));
 
-    if (options->family_id.has_value || options->extended_product_id.has_value)
+    if (options->enable_kss.value)
```
Comment (mingweishih) [inline tools/oesign/oesign.c:515]: Also unset the flag if EnableKSS is not specified. Isn't this tested by the prop_kss.c + sign.conf combination?

File: tests/config_id/enc/props_kss.c [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+
+#include <openenclave/enclave.h>
+
+OE_SET_ENCLAVE_SGX2(
+    1,    /* ProductID */
+    1,    /* SecurityVersion */
+    {0},  /* ExtendedProductID */
+    {0},  /* FamilyID */
+    true, /* Debug */
+    true, /* CapturePFGPExceptions */
```
Comment (mingweishih) [inline tests/config_id/enc/props_kss.c:12]: minor: setting `CapturePFGPExceptions` to false so that this tests focus only on the `RequireKSS`

File: tools/oesign/oesign.c (hunk 15 above)
Comment (manojrupireddy) [reply tools/oesign/oesign.c:515]: yes the test catches it. Forgot committing the change. Pushing now.

**Review by mingweishih — APPROVED** [review]
LGTM. Thanks for addressing all the comments!

**Comment by bodzhang** [comment]
> With the latest revisions, I have introduced a new field called 'EnableKSS' in conf file. User need to set this flag to 1 to leverage SGX KSS features. This will have indirect impact on developers planning to leverage FamilyID and ExtendedProductID fields of SGX2 machine. In the present code version of oesign.c in the master branch, kss flag is enabled if familyID and ExtendedProductID are present in conf file. With my current code, Instead of oesign.c deducting the logic to enable kss flag or not based on presence/absence of familyID and ExtendedProductId, developers need to explicitly mention if they want to enable KSS or not. This additional field will be helpful in the scenario where developers want to use configid/configsvn fields but not familyId and ExtendedProductiD fields. @bodzhang @yentsanglee @radhikaj @mingweishih @anakrish can you see if this new conf flag will have any unforeseen implications.

Outside of the familyID and ExtendedProductID context, configid/configsvn feature also has interaction with `sigstruct->attributemask.flags` and `sigstruct->attributes.flags` (set by the signing tool). I believe the original design goal is to enable a single signed SGX enclave binary with the ability to use configid/configsvn but can still function without it  to launch successfully on SGX CPUs without KSS support. With that goal, `sigstruct->attributemask.flags` and `sigstruct->attributes.flags` should both have SGX_FLAGS_KSS bit as 0, and the OE host side code will set secs.attributes' SGX_FLAGS_KSS bit based on whether the CPU support KSS and whether `context->config_data` is provided. 

The [existing implementation](https://github.com/openenclave/openenclave/blob/056f78ffc1bee2c8059a0efe98dbcfc05e5195f7/host/sgx/sgxsign.c#L662) of the signing tool seems to attempt to leave the SGX_FLAGS_KSS bit of `sigstruct->attributemask.flags` and `sigstruct->attributes.flags` as 0, if FamilyID and ExtendedProductID features are not used (requires KSS feature support, and unlike the configid/configsvn feature, can't be adjusted at enclave loading time). But the code didn't achieve the design goal, probably due to a bug. SGX_SIGSTRUCT_ATTRIBUTEMASK_FLAGS used to initialize sigstruct->attributemask.flags is defined as  0xfffffffffffffffb, already has SGX_FLAGS_KSS bit set. Fix the SGX_SIGSTRUCT_ATTRIBUTEMASK_FLAGS definition, changing it to 0xfffffffffffffff3, should make the implementation correct.

In summary, I believe the signing tool design is:

- If and only if developer opt-in with non-zero FamilyID or ExtendedProductID, the signing tool sets SGX_FLAGS_KSS bit of `sigstruct->attributemask.flags` and `sigstruct->attributes.flags` as 1;
- developer opt-in to configid/configsvn feature through `context->config_data` at run time, not at enclave signing time. The signing tool does not process configid/configsvn feature opt-in intention.

Therefore, the new 'EnableKSS' field in conf file is not needed.

**Comment by manojrupireddy** [comment]
Thank you @bodzhang . For clarification, I have a question.

- If OE host side code will set secs.attributes' SGX_FLAGS_KSS bit based on whether the CPU support KSS and whether context-config_data is provided, does it not affect the enclave measurement. Will the enclave measurement be same if host1 opts in for config_id feature and host2 opts out of config_id feature.

**Comment by bodzhang** [comment]
> Thank you @bodzhang . For clarification, I have a question.
> 
> * If OE host side code will set secs.attributes' SGX_FLAGS_KSS bit based on whether the CPU support KSS and whether context-config_data is provided, does it not affect the enclave measurement. Will the enclave measurement be same if host1 opts in for config_id feature and host2 opts out of config_id feature.

I had a typo in the previous post. The bitmask for SGX_FLAGS_KSS is 0x80, so the default SGX_SIGSTRUCT_ATTRIBUTEMASK_FLAGS value SGX_FLAGS_KSS bit as 0 should be 0xffffffffffffff**7b**.

Corrected the previous comment.

File: tests/config_id/enc/sign_kss.conf [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+
+# Enclave settings:
+Debug=1
+NumHeapPages=1024
+NumStackPages=1024
+NumTCS=2
+ProductID=1
+SecurityVersion=1
+FamilyID=47183823-2574-4bfd-b411-99ed177d3e43
+ExtendedProductID=2768c720-1e28-11eb-adc1-0242ac120002
```
Comment (mingweishih) [inline tests/config_id/enc/sign_kss.conf:12]: new line
